### PR TITLE
PSI: Cache file items

### PR DESCRIPTION
**Pull request by matklad**
Not sure if the implementation can be less copy-pasty...

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/psi/RsFile.kt [hunk 1]
```diff
@@ -70,50 +71,134 @@ class RsFile(
...
+
+    @Volatile
+    private var _items_cache: ItemsCache? = null
+
+    override fun subtreeChanged() {
+        super.subtreeChanged()
+        _items_cache = null
     }
+
+    private val items_cache: ItemsCache
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsFile.kt]: `_`

Comment (matklad) [reply src/main/kotlin/org/rust/lang/core/psi/RsFile.kt]: 😑

File: src/main/kotlin/org/rust/lang/core/psi/RsFile.kt [hunk 2]
```diff
@@ -70,50 +71,134 @@ class RsFile(
...
+
+    @Volatile
+    private var _itemsCache: ItemsCache? = null
+
+    override fun subtreeChanged() {
+        super.subtreeChanged()
+        _itemsCache = null
     }
+
+    private val itemsCache: ItemsCache
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/psi/RsFile.kt]: May be it's better to use `by lazy(PUBLICATION)` instead?

Comment (matklad) [reply src/main/kotlin/org/rust/lang/core/psi/RsFile.kt]: It won't work unfortunately because we need to clear cache manually in `subtreeChanged`.

**Pull request merged**